### PR TITLE
feature(driver): Nodejs 3.0 support websocket bulk pulling

### PR DESCRIPTION
**Pull request by xleili**

### CODE REVIEW

File: WebSocket/src/taosResult.ts [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+        // console.log(typeof taosResult.timing)
+        // console.log(typeof blocks.timing)
+        // console.log(blocks.id)
+
+        taosResult.timing = BigInt(taosResult.timing).valueOf() + blocks.timing;
+
+        const INT_32_SIZE = 4;
+
+        let bufferOffset = (4 * 5) + 8 + (4 + 1) * metaList.length;
```
Comment (sangshuduo) [inline WebSocket/src/taosResult.ts:96]: could you please add a few lines comments to describe what means about these numbers?

File: WebSocket/src/taosResult.ts [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+                    } else {
+                        colDataHead = colBlockHead + INT_32_SIZE * fetchResponse.rows + varOffset
+                        let dataLength = (new DataView(dataBuffer, colDataHead, 2).getInt16(0, true));
+                        // console.log("== loop var type length:" + dataLength)
+                        if (isVarType == ColumnsBlockType.VARCHAR) {
+
+                            row.push(readVarchar(dataBuffer, colDataHead + 2, dataLength));
+
+                        } else {
+                            row.push(readNchar(dataBuffer, colDataHead + 2, dataLength))
```
Comment (sangshuduo) [inline WebSocket/src/taosResult.ts:158]: some lines have semicolon, some do not

File: WebSocket/src/ut8Helper.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+const t3 = 224;
+const t4 = 240;
+const maskx = 63;
+const rune1Max = (1 << 7) - 1;
+
+const rune2Max = (1 << 11) - 1;
+const rune3Max = (1 << 16) - 1;
+
+
+// AppendRune appends the UTF-8 encoding of r to the end of p and
```
Comment (sangshuduo) [inline WebSocket/src/ut8Helper.ts:20]: AppendRun?

**Review by sangshuduo — APPROVED** [review]
LGTM with a few comments. not big deal but better to have a look

File: WebSocket/src/wsQueryInterface.ts [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        })
+    }
+
+    close() {
+        this._wsQueryClient.close();
+    }
+
+    checkURL(url: URL) {
+        // Assert is cloud url
+        if (!url.search.includes('?token=')) {
```
Comment (zitsen) [inline WebSocket/src/wsQueryInterface.ts:159]: check `&token=` or, to be better, save the query parameters in a map instead.

File: WebSocket/src/wsQueryInterface.ts [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    fetchBlock(fetchResponse: WSFetchResponse, taosResult: TaosResult): Promise<TaosResult> {
+        this._reqIDIncrement()
+        let fetchBlockMsg = {
+            action: 'fetch_block',
+            args: {
+                'req_id': this._req_id,
+                'id': fetchResponse.id,
+            }
+        }
+        return new Promise((resolve, reject) => {
```
Comment (zitsen) [inline WebSocket/src/wsQueryInterface.ts:108]: There're two kind of responses for fetchBlock:

1. binary data
2. json with error message like other actions.

File: WebSocket/example/basicUsage.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    `('2022-03-30 18:30:51.569',9,0,1,2,'binary3','nchar3')`;
+const querySql = `select * from ${db}.${table}`;
+const errorSql = 'show database';
+
+
+// ws.version()
+// .then((e) =>{console.log(e);ws.close()} );
+
+ws.connect()
+    .then(res=>console.log(res))
```
[the remark this answers is not in the document]
Comment (xleili) [reply WebSocket/example/basicUsage.ts:27]: Will submit in next pr  with a new example in 'async' and 'await' way.

File: WebSocket/src/wsQueryInterface.ts (hunk 5 above)
Comment (xleili) [reply WebSocket/src/wsQueryInterface.ts:108]: fix in next pr

**Pull request merged**